### PR TITLE
Add support keepSegmentGranularity for automatic compaction

### PR DESCRIPTION
**Pull request by jihoonson**
Fixes https://github.com/apache/incubator-druid/issues/6135.

### CODE REVIEW

File: docs/content/configuration/index.md [hunk 1]
```diff
@@ -782,9 +782,11 @@ A description of the compaction config is:
 |Property|Description|Required|
 |--------|-----------|--------|
 |`dataSource`|dataSource name to be compacted.|yes|
+|`keepSegmentGranularity`|Set [keepSegmentGranularity](../ingestion/compaction.html) to true for compactionTask.|no (default = true)|
 |`taskPriority`|[Priority](../ingestion/tasks.html#task-priorities) of compact task.|no (default = 25)|
-|`targetCompactionSizeBytes`|The target segment size of compaction. The actual size of a compact segment might be slightly larger or smaller than this value.|no (default = 838860800)|
-|`numTargetCompactionSegments`|Max number of segments to compact together.|no (default = 150)|
+|`inputSegmentSizeBytes`|Total input segment size of a compactionTask. The actual input size can be slightly larger than this value.|no (default = 419430400)|
```
Comment (fjy) [inline docs/content/configuration/index.md]: if the total size can be larger, what's the point of this config?

Comment (fjy) [reply docs/content/configuration/index.md]: It looks like this is a target size

File: server/src/main/java/org/apache/druid/server/coordinator/helper/SegmentCompactorUtil.java [hunk 2]
```diff
@@ -20,42 +20,30 @@
...
  */
 class SegmentCompactorUtil
 {
-  private static final Period LOOKUP_PERIOD = new Period("P1D");
-  private static final Duration LOOKUP_DURATION = LOOKUP_PERIOD.toStandardDuration();
   // Allow compaction of segments if totalSize(segments) <= remainingBytes * ALLOWED_MARGIN_OF_COMPACTION_SIZE
   private static final double ALLOWED_MARGIN_OF_COMPACTION_SIZE = .1;
 
-  static boolean isCompactible(long remainingBytes, long currentTotalBytes, long additionalBytes)
+  static boolean isCompactibleSize(long targetBytes, long currentTotalBytes, long additionalBytes)
```
Comment (fjy) [inline server/src/main/java/org/apache/druid/server/coordinator/helper/SegmentCompactorUtil.java]: this is a confusing name for a function - why rename?

File: server/src/main/java/org/apache/druid/server/coordinator/helper/NewestSegmentFirstIterator.java [hunk 3]
```diff
@@ -458,9 +401,14 @@ String getDataSource()
       this.segments = segments;
     }
 
-    public List<DataSegment> getSegments()
+    List<DataSegment> getSegments()
     {
       return segments;
     }
+
+    int size()
```
Comment (fjy) [inline server/src/main/java/org/apache/druid/server/coordinator/helper/NewestSegmentFirstIterator.java]: can we call this getSize() and explicitly list if it is public/private/protected?

File: server/src/main/java/org/apache/druid/server/coordinator/helper/NewestSegmentFirstIterator.java [hunk 4]
```diff
@@ -209,179 +234,97 @@ private void updateQueue(String dataSourceName, DataSourceCompactionConfig confi
...
-                config
+          // (*) Discard segments found so far because we can't compact them anyway.
+          final int numSegmentsToCompact = segmentsToCompact.size();
+          segmentsToCompact.clear();
+
+          if (!SegmentCompactorUtil.isCompactibleSize(inputSegmentSize, 0, timeChunkSizeBytes)) {
+            final DataSegment segment = chunks.get(0).getObject();
+            log.warn(
+                "shardSize[%d] for dataSource[%s] and interval[%s] is larger than inputSegmentSize[%d]."
+                + " Contitnue to the next shard.",
```
Comment (fjy) [inline server/src/main/java/org/apache/druid/server/coordinator/helper/NewestSegmentFirstIterator.java]: Continue*

File: server/src/main/java/org/apache/druid/server/coordinator/helper/SegmentCompactorUtil.java (hunk 2 above)
Comment (fjy) [reply server/src/main/java/org/apache/druid/server/coordinator/helper/SegmentCompactorUtil.java]: Oh I see, there's 2 functions, by Size and by Name

File: server/src/main/java/org/apache/druid/server/coordinator/helper/NewestSegmentFirstIterator.java (hunk 4 above)
Comment (jihoonson) [reply server/src/main/java/org/apache/druid/server/coordinator/helper/NewestSegmentFirstIterator.java]: Fixed. Thanks.

File: server/src/main/java/org/apache/druid/server/coordinator/helper/NewestSegmentFirstIterator.java (hunk 3 above)
Comment (jihoonson) [reply server/src/main/java/org/apache/druid/server/coordinator/helper/NewestSegmentFirstIterator.java]: Done.

File: docs/content/configuration/index.md (hunk 1 above)
Comment (jihoonson) [reply docs/content/configuration/index.md]: Hm makes sense. Fixed.

**Pull request merged**